Kernel/Bus/USB/UHCI/UHCIDescriptorPool.h

```diff
@@ -0,0 +1,90 @@
+// into the second template argument of AK::Stack. This class does NOT get to the point
+// where we will do dynamic allocation. Once you are out of descriptors, you'll receive
+// `nullptr` when you try to fetch one.
+template<typename T, size_t number_of_descriptors = 64>
+class UHCIDescriptorPool {
+
+    // Ensure that we can't get into a situation where we'll write past the page
+    // and blow up
+    static_assert(sizeof(T) * number_of_descriptors <= PAGE_SIZE);
+
+public:
+    static OwnPtr<UHCIDescriptorPool<T, number_of_descriptors>> try_create(const StringView name)
+    {
+        auto pool_memory_block = MM.allocate_kernel_region(PAGE_SIZE, "UHCI Descriptor Pool", Memory::Region::Access::ReadWrite);
+        if (!pool_memory_block)
+            return {};
+
+        return adopt_own_if_nonnull(new (nothrow) UHCIDescriptorPool(pool_memory_block.release_nonnull(), name));
+    }
+
+    UHCIDescriptorPool(NonnullOwnPtr<Memory::Region> pool_memory_block, const StringView& name)
+        : m_pool_name(name)
+        , m_pool_page(move(pool_memory_block))
+    {
+        // Clear out the pool page
+        memset(m_pool_page->vaddr().as_ptr(), 0, PAGE_SIZE);
+
+        // Go through the number of descriptors to create in the pool, and create a virtual/physical address mapping
+        for (size_t i = 0; i < number_of_descriptors; i++) {
+            auto placement_address = reinterpret_cast<void*>(m_pool_page->vaddr().get() + (i * sizeof(T)));
+            auto physical_address = static_cast<u32>(m_pool_page->physical_page(0)->paddr().get() + (i * sizeof(T)));
+            auto* object = new (placement_address) T(physical_address);
+            m_free_descriptor_stack.push(object); // Push the descriptor's pointer onto the free list
+        }
+    }
+
+public:
+    UHCIDescriptorPool() = delete;
+    ~UHCIDescriptorPool() = default;
+
+    [[nodiscard]] T* try_take_free_descriptor()
+    {
+        // We're out of descriptors!
+        if (m_free_descriptor_stack.is_empty())
+            return nullptr;
+
+        dbgln_if(UHCI_VERBOSE_DEBUG, "Got a free UHCI Descriptor @ {} from pool {}", m_free_descriptor_stack.top(), m_pool_name);
+        T* descriptor = m_free_descriptor_stack.top();
+        m_free_descriptor_stack.pop();
+
+        return descriptor;
+    }
+
+    void release_to_pool(T* ptr)
+    {
+        dbgln_if(UHCI_VERBOSE_DEBUG, "Returning descriptor @ {} to pool {}", ptr, m_pool_name);
+        if (!m_free_descriptor_stack.push(ptr))
+            dbgln("Failed to return descriptor to pool {}. Stack overflow!", m_pool_name);
+    }
+
+    void print_pool_information() const
+    {
+        dbgln("Pool {} allocated @ {}", m_pool_name, m_pool_page->physical_page(0)->paddr());
+    }
+
+private:
+    String m_pool_name;                                       // Name of this pool
+    NonnullOwnPtr<Memory::Region> m_pool_page;                // Physical page where descriptors actually reside
```

Comment:
This is not a "pool page", it's a region. Please make the name and comment reflect that.